include/nuttx/board.h

```diff
@@ -331,6 +331,28 @@ int board_uniquekey(FAR uint8_t *uniquekey);
 int board_switch_boot(FAR const char *system);
 #endif
 
+/****************************************************************************
+ * Name:  board_load_appimage
+ *
+ * Description:
+ *   Load an application image from a given load address and perform the
+ *   required actions for executing it (e.g. loading the Program Counter
+ *   register with the application image entry point address).
+ *
+ * Input Parameters:
+ *   loadaddr - Load address of the application image
+ *
+ * Returned Value:
+ *   If this function returns, then it was not possible to load the
+ *   application image due to some constraints. The return value in this case
+ *   is a board-specific reason for the failure.
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_BOARDCTL_LOAD_APPIMAGE
+int board_load_appimage(uintptr_t loadaddr);
```

Comment:
Look like this interface assume XIP(receive the loadaddr), how about the target image(from a raw partition or a file from file system) need to be loaded into RAM before running?

Reply:
Actually, this interface makes no assumption regarding whether the code will be executed in place or from RAM.
The `loadaddr` parameter refers to the base address of the application image on the flash storage.
The implementation of `board_load_appimage` may decide whether to copy the image from the load address into RAM, or to execute in place by simply overriding the PC with the application image entry point.

Reply:
Ok, understand. Here is my suggestion:

1. LOAD_APPIMAGE to LAUNCH_IMAGE or EXECUTE_IMAGE since it not only load but also execute image, and the image isn't an applicaiton but OS self.
2. It's better to change "uintptr_t loadaddr" to "FAR const char *image" or "FAR void *image". The address is too related to XIP, I could like to see the possibility to identify the image through the name not the address since the name match the file system more natually.

Reply:
[...] I could like to see the possibility to identify the image through the name not the address since the name match the file system more natually.

Just to make sure we are in sync: for *image* I am referring to a region in the Flash storage that contains a firmware image (code+data in RAW format), and not an update file.

Basically the idea is that the platform is expected to have some MTD partitions (e.g. `/dev/ota0` and `/dev/ota1`) that refer to booting slots.
So, initially, that `loadaddr` refers to the base address of either `/dev/ota0` partition or `/dev/ota1` partition.

Reply:
> > [...] I could like to see the possibility to identify the image through the name not the address since the name match the file system more natually.
> 
> Just to make sure we are in sync: for **image** I am referring to a region in the Flash storage that contains a firmware image (code+data in RAW format), and not an update file.
> 
> Basically the idea is that the platform is expected to have some MTD partitions (e.g. `/dev/ota0` and `/dev/ota1`) that refer to booting slots.
> Initially, that `loadaddr` refers to the base address of either `/dev/ota0` partition or `/dev/ota1` partition.

Yes, it point to a MTD partition in most case. But, why the path can't point to nuttx.bin on SD card? It's very useful in some case. It may even point to nuttx.elf and then board code can load the code/data to the correct location without hardcoding the magic address number in the source file.

> So, we may change the `loadaddr` parameter to be a path to MTD partition, but I feel that this may seen counter-intuitive, since the kernel will probably need to once again retrieve the base address of the selected partition.

The name(string) give many possibllity here: you can even pass itoa as the name here. The board code can:

1. If the string is actually number, it's an address
2. If the string is /dev/xxx, it's a partition
3. Otherwise, it's a file on the file system

Reply:
@xiaoxiang781216 Unfortunately sending the path as `const char *` parameter won't work out-of-the-box, because currently there is no interface for retrieving the starting address of an MTD partition.

To make things a bit more complicated, I've realized that besides the partition starting address (i.e `loadaddr` param) the application will also need to forward the size of the image header, which will be used as an offset for pointing to the beginning of the image's "useful" part (code + data). So instead of a single value, we'd need to provide a struct, something like:

```c
struct boot_info_s
{
  uint32_t img_header_size;   /* Size of the image header in bytes */
  uint32_t img_addr;               /* MTD partition starting address */
};
```

If we create this MTD interface for retrieving the partition starting address, `img_addr` may be easily substituted by a `const char * img_path` field.
What's your suggestion on this?

Reply:
> @xiaoxiang781216 Unfortunately sending the path as `const char *` parameter won't work out-of-the-box, because currently there is no interface for retrieving the starting address of an MTD partition.
> 
> To make things a bit more complicated, I've realized that besides the partition starting address (i.e `loadaddr` param) the application will also need to forward the size of the image header, which will be used as an offset for pointing to the beginning of the image's "useful" part (code + data).

If you want to write a general bootloader, this detail information(e.g. image header) need be handled by board specific code since the different vendor may define a different image header.

> So instead of a single value, we'd need to provide a struct, something like:
> 
> ```c
> struct boot_info_s
> {
>   uint32_t img_header_size;   /* Size of the image header in bytes */
>   uint32_t img_addr;          /* MTD partition starting address */
> };
> ```
> 
> If we create this MTD interface for retrieving the partition starting address, `img_addr` may be easily substituted by a `const char * img_path` field.
> What's your suggestion on this?

It's general question to get the block device offset and size, we can reference how BSD or Linux define the IOCTL. At least FIOC_MMAP/MTDIOC_XIPBASE can answer some query.

Reply:
BTW, I've updated the PR with another proposal for addressing the naming issue. I've changed *load* to simply **boot**, which is more intuitive for the intended purpose.

Reply:
I've changed `board_boot_image` to receive a string path instead of the address.

> It's general question to get the block device offset and size, we can reference how BSD or Linux define the IOCTL. At least FIOC_MMAP/MTDIOC_XIPBASE can answer some query.

See https://github.com/apache/incubator-nuttx/pull/4248, there I've proposed a new IOCTL for addressing this requirement.